book/invalidation.md

```diff
@@ -12,6 +12,118 @@ time in layout, not graphics. Yet just like compositing enabled smooth
+
+Now, in the case of typing into an `input` element, our browser
+rebuilds the layout tree, but in fact the layout tree looks the same
+each time---the text inside an `input` element doesn't get its own
+layout object. But browsers support other forms of text editing which
+does affect the layout tree. For example, in a real browser you can
+give any element the `contenteditable` attribute to make it editable.
+Once you do so, the user can click on the element to add or modify
+text in it:
+
+::: {.example contenteditable=true}
+Click on this text to edit it.
+:::
+
+Let's implement this in our browser---it will make a good test of
+invalidation. To begin with, we need to make elements with a
+`contenteditable` property focusable:
+
+``` {.python}
+def is_focusable(node):
+    # ...
+    elif "contenteditable" in node.attributes:
+        return True
+    # ...
+```
+
+Once we're focused on an editable node, key presses should add text to
+it. In a real browser, the implementation would be a bit more complex
+in order to handle cursor movement, but I'm going to have key presses
+append to the last text node in the editable element. First we need to
+find that element:
+
+``` {.python}
+class Frame:
+    def keypress(self, char):
+        # ...
+        elif self.tab.focus and "contenteditable" in self.tab.focus.attributes:
+            text_nodes = [
+               t for t in tree_to_list(self.tab.focus, [])
+               if isinstance(t, Text)
+            ]
+            if text_nodes:
+                last_text = text_nodes[-1]
+            else:
+                last_text = Text("", self.tab.focus)
+                self.tab.focus.children.append(last_text)
+```
+
+Note that if the editable element has no text children I append a new
+one. Now we need to add the typed character to this element:
+
+``` {.python}
+class Frame:
+    def keypress(self, char):
+        elif self.tab.focus and "contenteditable" in self.tab.focus.attributes:
+            # ...
+            last_text.text += char
+            self.set_needs_render()
+```
+
+This is enough to make editing work, but it's convenient to also draw
+a cursor to show the user where the editing will happen and to provide
+visual confirmation that they've focused on the right element. Let's
+do that in `BlockLayout`:
+
+``` {.python}
+class BlockLayout:
+    def paint(self, display_list):
+        # ...
+        if self.node.is_focused and "contenteditable" in self.node.attributes:
+            text_nodes = [
+                t for t in tree_to_list(self, [])
+                if isinstance(t, TextLayout)
+            ]
+            if text_nodes:
+                cx = text_nodes[-1].x + text_nodes[-1].width
+                cy1 = text_nodes[-1].y
+                cy2 = text_nodes[-1].y + text_nodes[-1].height
+            else:
+                cx = self.x
+                cy1 = self.y
+                cy2 = self.y + self.height
+            cmds.append(DrawLine(cx, cy1, cx, cy2))
```

Comment:
Factor out similar code for `InputLayout` and call it from each place, to draw the reader's
attention to the shared concept?